### ISSUE TITLE
fft_small build fails on arm64

### ISSUE THREAD

**Issue by math-gout** [flintlib/flint2#1361]
I tried compiling on a M1 Macbook but ```src/fft_small/nmod_poly_mul.c``` wants to use the following functions which are not implemented for NEON and causes the build to fail :
```
vec8d_set_d
vec8n_addmod
vec8n_addmod_limited
vec8n_bit_and
vec8n_bit_shift_right
vec8n_convert_limited_vec8d
vec8n_load_unaligned
vec8n_set_n
```

EDIT: I also tried compiling on a Raspberry Pi 4 and got the same issue but this time ```gcc``` is also complaining about ```__builtin_addcl``` and ```__builtin_subcl``` being undefined in ```src/crt_helpers.h```

**Comment by math-gout** [comment]
@fredrik-johansson on your blog you talked about a ```--enable-neon``` configuration option, but when I run ```./configure --help``` I see no such option.

**Comment by fredrik-johansson** [comment]
I assume that mpn_mul.c compiles though? From the looks of it Dan didn't finish NEON versions of those SIMD functions needed by nmod_poly_mul. It shouldn't be too hard to implement them, but we'll need CI for aarch64 to be sure things work.

You're right that the config flag hasn't been added yet.

**Comment by math-gout** [comment]
> I assume that mpn_mul.c compiles though?

Yes it does, I have a mpn_mul.o file.

**Comment by fredrik-johansson** [comment]
I've tried to get the arm64 build working again on this branch https://github.com/flintlib/flint2/tree/arm64

I have temporarily disabled the ``nmod_poly_mul`` code as there are some issues with the ``mpn_mul`` code that need to be sorted out first.

Specifically, in ``crt_helpers.h``, I have disabled the inline assembly versions of the ``multi_add`` / ``multi_sub`` functions as ``fft_small/test/t-mul`` hangs in ``_reduce_big_sum`` when these are used. I have no idea why this hang occurs. (Compiler bug? I have gcc 12.2.0 on the arm64 machine.)

Can you check if this branch builds, and if so, can you check if ``fft_small/test/t-mul`` passes?

And if it does, can you check what happens if you re-enable the inline assembly in ``crt_helpers.h`` (change the ``#if 0`` on line 85 to ``#if 1``)?

**Comment by math-gout** [comment]
This branch now builds and ```fft_small/test/t-mul``` passed when I ran ```make check MOD="fft_small"```, but I now have the following error :

```
In file included from src/fft_small/test/t-nmod_poly_mul.c:4:
In file included from ./src/fft_small.h:15:
Undefined symbols for architecture arm64:
  "__nmod_poly_divrem_precomp", referenced from:
      _main in t-nmod_poly_mul-67cb0b.o
  "__nmod_poly_divrem_precomp_init", referenced from:
      _main in t-nmod_poly_mul-67cb0b.o
  "__nmod_poly_mul_mid_mpn_ctx", referenced from:
      _main in t-nmod_poly_mul-67cb0b.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [build/fft_small/test/t-nmod_poly_mul] Error 1
```

I tried to re-enable the inline assembly in ```crt_helper.h``` anyway, and it still builds, but ```fft_small/test/t-mul``` now fails.

**Comment by fredrik-johansson** [comment]
Thanks, that's helpful; then it's not just the machine I'm testing on that is broken.

**Comment by tthsqe12** [comment]
Two things (it is probably the second)
1. at some point the input arguments to subborrow_u64 got reversed in gcc
 2. there are missing "&"s on some of the outputs in sub_dddmmmmmsssssss and add_ßàaa in the arm section. As I recall, this & means that the register is not supposed to be the same (?) as any other register in the asm clause. This is important so that the writing of the outputs doesn't clobber the inputs. As such, all but the last output reg should have a &. Of course this is overkill since it is fine if the second to last output reg is the same as either of the first two input regs, for example, but there is no way to communicate this fact to the c compiler. I think the & are done correctly in the amd64 section.

I would say check the output of the c fallback, and if the compiler isn't being as stupid as gcc was on the gpr-starved amd64 target, scrap the add_ssssaaaa nonsense (which has gone too far) and go with the c fallback using intrinsics in a loop.

I can help more when I get laptop+wifi.

**Comment by fredrik-johansson** [comment]
I believe I found the problem: besides missing "&"s, some ``adc``/``sbc`` instructions should be ``adcs``/``sbcs``. I will fix and add some test code.

It's too bad GCC doesn't have ``__builtin_addcl`` and I doubt ``__builtin_add_overflow`` does the right thing. What about using ``uint128``? Indeed, the ``_mul`` and ``_madd`` functions look equivalent to ``umul_ppmm`` and ``add_ssaaaa``, but you must have done them this way because it results in faster code. In that case, shouldn't we change the implementations of ``umul_ppmm`` and ``add_ssaaaa``?

**Comment by fredrik-johansson** [comment]
Tests now seem to pass on the arm64 branch. On to fixing the polynomials...

**Comment by fredrik-johansson** [comment]
I implemented the missing vector functions and tests passed on the first try!

It's missing optimized versions of ``addmod`` and ``addmod_limited`` though.

**Comment by fredrik-johansson** [comment]
Potentially fixed now, but needs testing.

**Comment by math-gout** [comment]
Just tested it on a Raspberry Pi 4 and it builds and all tests are passing.

I also tested on an M1 Macbook but this time it won't build and spits out this error :
```
In file included from src/mpn_extras/mul_large.c:28:
In file included from ./src/fft_small.h:15:
./src/machine_vectors.h:1487:12: error: argument to '__builtin_neon_vshrq_n_v' must be a constant integer
    return vshrq_n_u64(a, n);
           ^              ~
```

I tried fixing it by replacing with line with :
```c
return vshlq_s64(a, vdupq_n_s64(-n));
```
But this doesn't fix the problem as I now get this error when running tests :
```
nmod_poly_mul....divrem nbits:  2Flint exception (General error):
    Unable to allocate 2048 bytes with alignment 4096 in flint_aligned_alloc
make: *** [build/fft_small/test/t-nmod_poly_mul_TEST_RUN] Abort trap: 6
```

**Comment by fredrik-johansson** [comment]
It's possible that your solution is fine though; the test failure you're seeing might be unrelated.

**Comment by fredrik-johansson** [comment]
This code looks suspicious:

    stride = n_round_up(sd_fft_ctx_data_size(depth), 128);
    ...
    M->bbuf = flint_aligned_alloc(4096, np*stride*sizeof(double));

@tthsqe12 is the second line meant to be

    M->bbuf = flint_aligned_alloc(4096, n_round_up(np*stride*sizeof(double), 4096));

?

**Comment by tthsqe12** [comment]
ah, this is mallocing the buffer for precomp init.  Yeah, the alignment has to be at least 32, i would suggest at least 64, and i think the malloc size has to be a multiple of the alignment

**Comment by tthsqe12** [comment]
it seems yall are taking this seriously. Can you check if the performance degration from removing the shift, i.e. 0* instead of 4* here
https://github.com/flintlib/flint2/blob/trunk/src/fft_small.h#L205
is acceptable? If so, I've got some performance enhancements and interface simplifications you might like. No one should be thinking about BLK_SZ this much.

**Comment by fredrik-johansson** [comment]
I'm not seeing any performance degradation on macro benchmarks.

I'm all in favor of making that simplification.

**Comment by fredrik-johansson** [comment]
Just pushed the possible fix for machine_vectors issue and for the aligned alloc.

@math-gout Can you try again?

**Comment by math-gout** [comment]
Still no luck
```
fmpz_poly_mul....PASS
mpn_add_inplace_c....PASS
mpn_mul....PASS 5000/5000
nmod_poly_mul....sqr nbits: 64(squaring) mulmid error at index 1757
zl=1757, zh=2599, an=3372
mod: 14194180165020616007
make: *** [build/fft_small/test/t-nmod_poly_mul_TEST_RUN] Abort trap: 6
```

**Comment by fredrik-johansson** [comment]
Is this on the M1? The Raspberry Pi 4 still works?

I regret that I don't have an M1 myself to debug this...

**Comment by math-gout** [comment]
This is on the M1, i'll check on the Raspberry Pi 4 too.

**Comment by math-gout** [comment]
The ```fft_small``` tests pass on the Raspberry Pi, but I had to change ```vshlq_s64``` to ```vshlq_u64``` because it wouldn't compile otherwise.

I also tried with this change on the M1 and this time got : 
```
fmpz_poly_mul....PASS
mpn_add_inplace_c....PASS
mpn_mul....PASS 5000/5000
nmod_poly_mul....divrem nbits: 10t-nmod_poly_mul(29926,0x1ebc85b40) malloc: Incorrect checksum for freed object 0x102511a00: probably modified after being freed.
Corrupt value: 0x34a
t-nmod_poly_mul(29926,0x1ebc85b40) malloc: *** set a breakpoint in malloc_error_break to debug
make: *** [build/fft_small/test/t-nmod_poly_mul_TEST_RUN] Abort trap: 6
```

**Comment by tthsqe12** [comment]
If you want to valgrind this on x86 and work around the valgrind sign bit bug, you'll have to sanitise the last input to blendv by making sure it is the (mask) result of a proper comparison with zero.
that is, when c is not already the result of a comparison
blendv(a, b, c) -> blendv(a,b,c<0)
note the result of a comparison in an fp register is not a real fp value, it's a mask of either all zeros or all ones.
note also that the hardware blendv only uses the top bit of c, which is why the current code works and why valgrind breaks it.
no idea about fp comparisons on aarch64

**Comment by jpflori** [comment]
On a Mac M1 I reach:
`  CC  nmod_poly/test/t-exp_series
exp_series....FAIL:
...
`

**Comment by jpflori** [comment]
I did not pass any specific flags to the configure script yet.

**Comment by jpflori** [comment]
And I'm on trunk.

**Comment by jpflori** [comment]
On arm64 branch I also add to apply @math-gout diff to compile:

```
diff --git a/src/machine_vectors.h b/src/machine_vectors.h
index 132b49b8c..42de08fdc 100644
--- a/src/machine_vectors.h
+++ b/src/machine_vectors.h
@@ -1484,7 +1484,7 @@ EXTEND_VEC_DEF2(vec4n, vec8n, _bit_and)
 
 FLINT_FORCE_INLINE vec2n vec2n_bit_shift_right(vec2n a, ulong n)
 {
-    return vshrq_n_u64(a, n);
+    return vshlq_s64(a, vdupq_n_s64(-n));
 }
 
 FLINT_FORCE_INLINE vec4n vec4n_bit_shift_right(vec4n a, ulong n)
```

**Comment by jpflori** [comment]
In both `trunk` and `arm64` branches I get many warnings like
```
./src/machine_vectors.h:1415:22: warning: incompatible pointer types assigning to 'const unsigned long long *' from 'const mp_limb_t *' (aka 'const unsigned long *') [-Wincompatible-pointer-types]
    return vld1q_u64(a);
```

**Comment by jpflori** [comment]
My `make check` on `arm64` also fails at `exp_series....FAIL:`

**Comment by jpflori** [comment]
> Just pushed the possible fix for machine_vectors issue and for the aligned alloc.
> 
> @math-gout Can you try again?

BTW I do not see these commits on the `arm64` branch.

**Comment by jpflori** [comment]
I also get a `malloc/free` issue when running `make check MOD="fft_small"`

**Comment by fredrik-johansson** [comment]
> BTW I do not see these commits on the arm64 branch.

Sorry, these were pushed to ``main`` but not the ``arm64`` branch. It should be rebased now.

**Comment by jpflori** [comment]
Looks like _nmod_poly_reverse called from _nmod_poly_divrem_mpn_ctx writes outside of a mallocated memory (from _nmod_poly_divrem_mpn_ctx).

**Comment by fredrik-johansson** [comment]
Indeed, it looks like one can have ``bn > N`` in ``_nmod_poly_divrem_mpn_ctx``.

I guess the problem is that ``2^N`` is chosen large enough for ``t`` to hold ``bn-1`` coefficients, but then ``t`` gets used as scratch space for reversing a length ``bn`` polynomial.

It looks like changing

    ulong* t = FLINT_ARRAY_ALLOC(N, ulong);

to

    ulong* t = FLINT_ARRAY_ALLOC(FLINT_MAX(N, bn), ulong);

could fix it.

**Comment by jpflori** [comment]
Now I get
```
nmod_poly_mul....divrem nbits: 64quotient error at index 2435
qn=2436, an=2604, bn=169
mod: 13857350307591655756
```

**Comment by fredrik-johansson** [comment]
OK, the error occurring only for nbits = 64 makes me suspect some signed/unsigned issue in machine_vectors.h or a shift by 64 bits or something like that. Maybe the warnings (if you still have them) can give a clue.

It is curious however that this error is first reported for the division code.

If you change the test program to run a lot of iterations for polynomial multiplication with 64-bit modulus, is it possible to trigger a test failure?

**Comment by jpflori** [comment]
Indeed:
```
nmod_poly_mul....mul nbits: 64mulmid error at index 405
zl=405, zh=1463, an=5066, bn=35
mod: 16458071311883748506
```

**Comment by jpflori** [comment]
Warnings are:
```
In file included from ./src/fft_small.h:15:
./src/machine_vectors.h:829:36: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
FLINT_FORCE_INLINE vec1d vec1d_zero() {
                                   ^
                                    void
./src/machine_vectors.h:833:35: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
FLINT_FORCE_INLINE vec1d vec1d_one() {
                                  ^
                                   void
./src/machine_vectors.h:961:36: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
FLINT_FORCE_INLINE vec2d vec2d_zero()
                                   ^
                                    void
./src/machine_vectors.h:966:35: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
FLINT_FORCE_INLINE vec2d vec2d_one()
                                  ^
                                   void
./src/machine_vectors.h:1239:1: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
EXTEND_VEC_DEF0(vec2d, vec4d, _zero)
^
./src/machine_vectors.h:756:26: note: expanded from macro 'EXTEND_VEC_DEF0'
FLINT_FORCE_INLINE V V##f() { \
                         ^
./src/machine_vectors.h:1240:1: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
EXTEND_VEC_DEF0(vec2d, vec4d, _one)
^
./src/machine_vectors.h:756:26: note: expanded from macro 'EXTEND_VEC_DEF0'
FLINT_FORCE_INLINE V V##f() { \
                         ^
./src/machine_vectors.h:1325:1: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
EXTEND_VEC_DEF0(vec4d, vec8d, _zero)
^
./src/machine_vectors.h:756:26: note: expanded from macro 'EXTEND_VEC_DEF0'
FLINT_FORCE_INLINE V V##f() { \
                         ^
./src/machine_vectors.h:1326:1: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
EXTEND_VEC_DEF0(vec4d, vec8d, _one)
^
./src/machine_vectors.h:756:26: note: expanded from macro 'EXTEND_VEC_DEF0'
FLINT_FORCE_INLINE V V##f() { \
                         ^
./src/machine_vectors.h:1373:14: warning: incompatible pointer types assigning to 'unsigned long long *' from 'mp_limb_t *' (aka 'unsigned long *') [-Wincompatible-pointer-types]
   vst1q_u64(z, a);
             ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/14.0.3/include/arm_neon.h:26801:26: note: expanded from macro 'vst1q_u64'
  __builtin_neon_vst1q_v(__p0, (int8x16_t)__s1, 51); \
                         ^~~~
In file included from src/fft_small/test/t-nmod_poly_mul.c:4:
In file included from ./src/fft_small.h:15:
./src/machine_vectors.h:1415:22: warning: incompatible pointer types assigning to 'const unsigned long long *' from 'const mp_limb_t *' (aka 'const unsigned long *') [-Wincompatible-pointer-types]
    return vld1q_u64(a);
                     ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/14.0.3/include/arm_neon.h:8741:47: note: expanded from macro 'vld1q_u64'
  __ret = (uint64x2_t) __builtin_neon_vld1q_v(__p0, 51); \
```

**Comment by jpflori** [comment]
Another failing example:
```
nmod_poly_mul....mul nbits: 64mulmid error at index 2097
zl=2097, zh=2980, an=5756, bn=34
mod: 14442458465753067646
```

**Comment by fredrik-johansson** [comment]
Probably the error can be triggered with a much shorter input (length 1 might be enough).

**Comment by jpflori** [comment]
It also happens as soon as bn=2 when an=1
```
nmod_poly_mul....mul nbits: 64mulmid error at index 1
zl=0, zh=8, an=1, bn=2
mod: 11326125079979797344
```
Not sure what zl and zh mean.

**Comment by jpflori** [comment]
Does it make sense to have zh > an + bn - 1 ?

**Comment by fredrik-johansson** [comment]
> Not sure what zl and zh mean.

I guess you already figured it out, but: lower and higher bounds for mulmid as documented (very tersely) on https://www.flintlib.org/doc/fft_small.html. In particular zl = 0 does a mullow of length zh.

> Does it make sense to have zh > an + bn - 1 ?

Sure, in cases where you want to zero-pad the result. Traditionally we don't allow this in _nmod_poly_mullow etc., requiring the caller to take care of such cases. I guess for mulmid, which has more input combinations, it is useful to deal with this case so that there are fewer possible ways for the caller to provide incorrect input.

**Comment by fredrik-johansson** [comment]
> It also happens as soon as bn=2 when an=1

I guess at this point you could just trace through the multiplication and see where it does the wrong thing. You have (1) reduction mod two different p:s, (2) FFTs, pointwise muls and IFFT, and (3) CRT recombination from the two mod p results; most likely the error is in (1) or (3). If you have an x86-64 machine, you could do the same multiplication there and check where the values diverge.

(I'm still waiting for access to a machine where I can try this.)

**Comment by jpflori** [comment]
Should not be very hard to debug:
```
nmod_poly_mul....mul nbits: 64
a 5070468039402751120
b 2211287077698730482
c d
1 0 15221983403907729074
2 0 7864103106316499422
3 0 6860885197721404584
4 0 0
5 0 0
6 0 0
7 0 0
8 0 0
9 0 0
10 0 0
mulmid error at index 1
zl=1, zh=11, an=1, bn=4
mod: 15581639867842951658
```

**Comment by jpflori** [comment]
looks like one of the issues is that _nmod_poly_mul gets called with len1 < len2.

**Comment by jpflori** [comment]
Arf that it "just" the call in the test code...

**Comment by fredrik-johansson** [comment]
Despite what the documentation says, ``len1 < len2`` *should* be fine here. But of course it's worth checking that that isn't the issue...

**Comment by jpflori** [comment]
In _nmod_poly_mul_classical, there is
    if (len1 == 1)
    {
        res[0] = nmod_mul(poly1[0], poly2[0], mod);
        return;
    }
which assumes len1 >= len2

**Comment by jpflori** [comment]
In my branch I have
```
            _nmod_poly_mul(c, a, an, b, bn, mod);
            _nmod_poly_mul_mid_mpn_ctx(d, zl, zh, a, an, b, bn, mod, R);
```

**Comment by jpflori** [comment]
I guess I need to git pull...

**Comment by jpflori** [comment]
With the latest commits, using _nmod_poly_mul_KS, I only get an error in squaring when an >= 256.

**Comment by jpflori** [comment]
And output of _nmod_poly_mul_mid_mpn_ctx is only made of zeros.

**Comment by jpflori** [comment]
This mitigates the issue for sqr and mul but not for divrem:
```
diff --git a/src/fft_small/nmod_poly_mul.c b/src/fft_small/nmod_poly_mul.c
index 15d1c6211..2443131cd 100644
--- a/src/fft_small/nmod_poly_mul.c
+++ b/src/fft_small/nmod_poly_mul.c
@@ -191,7 +191,7 @@ FLINT_ASSERT(i+j < atrunc);
         for (i = 0; i + BLK_SZ <= an; i += BLK_SZ)
         {
             aI = sd_fft_ctx_blk_index(abuf, i/BLK_SZ);
-#if 1
+#if 0
             for (j = 0; j < BLK_SZ; j += 8)
             {
                 vec8n t = vec8n_load_unaligned(a + i + j);
```

**Comment by jpflori** [comment]
I guess vec8n_bit_shift_right(t, 32); copies the sign/most significant bit.

**Comment by jpflori** [comment]
Then vec8n_convert_limited_vec8d produces a NaN.

**Comment by jpflori** [comment]
> ```c
> )
> ```

Which is caused by the above.

**Comment by jpflori** [comment]
This helps but is not sufficient:
```
diff --git a/src/machine_vectors.h b/src/machine_vectors.h
index b510f0faa..c4a0fc19f 100644
--- a/src/machine_vectors.h
+++ b/src/machine_vectors.h
@@ -1502,6 +1502,26 @@ FLINT_FORCE_INLINE vec8n vec8n_bit_shift_right(vec8n a, ulong n)
     return z;
 }
 
+FLINT_FORCE_INLINE vec2n vec2n_bit_shift_right_32(vec2n a)
+{
+    /* vshrq_n_u64(a, n) cannot be used because n must be a compile-time
+       constant, and the compiler doesn't see that n is constant
+       even if the function is forced inline */
+    return vshrq_n_u64(a, 32);
+}
+
+FLINT_FORCE_INLINE vec4n vec4n_bit_shift_right_32(vec4n a)
+{
+    vec4n z = {vec2n_bit_shift_right_32(a.e1), vec2n_bit_shift_right_32(a.e2)};
+    return z;
+}
+
+FLINT_FORCE_INLINE vec8n vec8n_bit_shift_right_32(vec8n a)
+{
+    vec8n z = {vec4n_bit_shift_right_32(a.e1), vec4n_bit_shift_right_32(a.e2)};
+    return z;
+}
+
 
 #undef EXTEND_VEC_DEF4
 #undef EXTEND_VEC_DEF3
```

**Comment by jpflori** [comment]
vec8d_mulmod is problematic

**Comment by fredrik-johansson** [comment]
> vec8d_mulmod is problematic

What about it? It looks functionally identical to the AVX version.

**Comment by jpflori** [comment]
Result is wrong, did not look at each step of the mulmod yet.

**Comment by jpflori** [comment]
Good news: issue in divrem was just because there was another instance of `vec8n_bit_shift_right_32(t, 32)` in `_mod_red`.

**Comment by tthsqe12** [comment]
Just dropping in here to see if everything is going. The original code was not completed for aarch64, as is now painfully clear. It looks like a bad compiler has prompted the replacement of unsigned shifts with signed ones and has of course caused problems with 64  bit values.

1. It seems from the discussion here that the right bit shift needs the shift amount baked into the instruction, and thus the compiler wants this immediate integer at compile time. If the compiler is too primitive to figure out that this is precisely what we are giving it (32), then it is indeed a good idea to put the _32 as a template argument.
2. One cannot replace unsigned right shits by signed right shits.
3. The return range of mulmod is complicated and explained in the readme.

**Comment by jpflori** [comment]
Solution is indeed to define macros for unsigned right shifts with constant values.
Not though that only the 32 value seem to be used in the code if nobody wants to do the templating.
Force inline does not seem to work. Not sure the compilers are so primitive, they are the latest versions of clang and gcc... Maybe it is worth an upstream ticket.

**Comment by fredrik-johansson** [comment]
@jpflori Excellent. Could you prepare a patch or PR that fixes this issue and the other machine_vectors warnings?

It's fine to have methods for the constant shift by 32; no need to introduce templates until someone needs them.

**Comment by jpflori** [comment]
[shift32.patch](https://github.com/flintlib/flint2/files/11461323/shift32.patch)

**Comment by fredrik-johansson** [comment]
Thanks! Do the tests pass on the ``trunk`` branch now?

**Comment by tthsqe12** [comment]
eh, what about a 64 case switch statement in the original code? I know it looks bad, but another layer of C templating looks even worse. Once you have _32, that implies .. , _31, _32, _33, ..., and this is a template, regardless of how many cases are actually implemented.
This is clearly a shortcoming in the implementation of these intrinsics not present in the x86 implementation.

**Comment by fredrik-johansson** [comment]
Great. Thanks for the help!
